src/ray/raylet/reconstruction_policy.h

```diff
@@ -2,33 +2,123 @@
 
-  /// Check whether an object requires reconstruction. If this object requires
-  /// reconstruction, the registered task reconstruction handler will be called
-  /// for each task that needs to be re-executed.
+  /// Listen for task lease notifications about an object that may require
+  /// reconstruction. If no notifications are received within the initial
+  /// timeout, then the registered task reconstruction handler will be called
+  /// for the task that created the object.
   ///
   /// \param object_id The object to check for reconstruction.
-  void CheckObjectReconstruction(const ObjectID &object_id);
+  void Listen(const ObjectID &object_id);
+
+  /// Cancel listening for an object. Notifications for the object will be
+  /// ignored. This does not cancel a reconstruction attempt that is already in
+  /// progress.
+  ///
+  /// \param object_id The object to cancel.
+  void Cancel(const ObjectID &object_id);
+
+  /// Handle a notification for a task lease. This handler should be called to
+  /// indicate that a task currently being executed, so any objects that it
+  /// creates should not be reconstructed.
+  ///
+  /// \param task_id The task ID of the task being executed.
+  /// \param expires_at The time at which the task's lease expires. If a second
+  /// notification is not received within this timeout, then objects that the
+  /// task creates may be reconstructed.
+  void HandleTaskLeaseNotification(const TaskID &task_id, int64_t expires_at);
 
  private:
+  struct ReconstructionTask {
+    ReconstructionTask(boost::asio::io_service &io_service)
+        : created_objects(),
+          expires_at(INT64_MAX),
+          subscribed(false),
+          reconstruction_attempt(0),
+          reconstruction_timer(new boost::asio::deadline_timer(io_service)) {}
+
+    // The objects created by this task that we are listening for notifications for.
+    std::unordered_set<ObjectID> created_objects;
+    // The time at which the current lease for this task expires.
+    int64_t expires_at;
+    // Whether we are subscribed to lease notifications for this task.
+    bool subscribed;
+    // The number of times we've attempted reconstructing this task so far.
+    int reconstruction_attempt;
+    // The task's reconstruction timer. If this expires before a lease
+    // notification is received, then the task will be reconstructed.
+    std::unique_ptr<boost::asio::deadline_timer> reconstruction_timer;
+  };
+
+  // Set the reconstruction timer for a task. If no task lease notifications
+  // are received within the timeout, then reconstruction will be triggered.
+  void SetTaskTimeout(std::unordered_map<TaskID, ReconstructionTask>::iterator task_it,
+                      int64_t timeout_ms);
+
+  /// Reconstruct a task.
+  void AttemptReconstruction(const TaskID &task_id, const ObjectID &object_id,
+                             int reconstruction_attempt);
```

Comment:
The `reconstruction_attempt` argument is used for suppression, right? Can you explain this in the comment?

Reply:
Yep!